### PR TITLE
Implement FVM ValidatePublicKey

### PR DESCRIPTION
**Pull request by janezpodhostnik**
ref: https://github.com/dapperlabs/flow-go/issues/5541

### CODE REVIEW

File: fvm/crypto/crypto.go [hunk 1]
```diff
@@ -150,6 +150,22 @@ func CryptoToRuntimeHashingAlgorithm(h hash.HashingAlgorithm) runtime.HashAlgori
...
 }
 
+// ValidatePublicKey returns true if public key is valid
+func ValidatePublicKey(signAlgo runtime.SignatureAlgorithm, pk []byte) (valid bool, err error) {
+	sigAlgo := RuntimeToCryptoSigningAlgorithm(signAlgo)
+
+	_, err = crypto.DecodePublicKey(sigAlgo, pk)
+
+	if err != nil {
+		// TODO: `crypto.IsInvalidInputsError(err)` doesn't work here!
```
Comment (tarakby) [inline fvm/crypto/crypto.go:160]: What was the issue with using `crypto.IsInvalidInputsError(err)` ?

File: fvm/crypto/crypto_test.go [hunk 2]
```diff
@@ -227,3 +227,53 @@ func TestVerifySignatureFromRuntime(t *testing.T) {
...
+		return pk.PublicKey().Encode()
+	}
+
+	t.Run("Unknown algorithm should return false", func(t *testing.T) {
+		valid, err := crypto.ValidatePublicKey(runtime.SignatureAlgorithmUnknown, validPublicKey(t, runtime.SignatureAlgorithmECDSA_P256))
+		require.NoError(t, err)
+		require.False(t, valid)
+	})
+
+	t.Run("valid public key should return true", func(t *testing.T) {
```
Comment (tarakby) [inline fvm/crypto/crypto_test.go:247]: 👌

File: fvm/crypto/crypto.go [hunk 3]
```diff
@@ -150,6 +150,22 @@ func CryptoToRuntimeHashingAlgorithm(h hash.HashingAlgorithm) runtime.HashAlgori
...
+		if _, ok := err.(*crypto.InvalidInputsError); ok {
+			return false, nil
+		}
+		return false, err
```
Comment (tarakby) [inline fvm/crypto/crypto.go:164]: maybe we could wrap the error:
```suggestion
		return false, fmt.Errorf("validate public key failed: %w", err)
```

File: fvm/crypto/crypto_test.go [hunk 4]
```diff
@@ -227,3 +227,53 @@ func TestVerifySignatureFromRuntime(t *testing.T) {
...
+	t.Run("invalid public key should return false", func(t *testing.T) {
+		signatureAlgos := []runtime.SignatureAlgorithm{
+			runtime.SignatureAlgorithmECDSA_P256,
+			runtime.SignatureAlgorithmECDSA_secp256k1,
+			runtime.SignatureAlgorithmBLS_BLS12_381,
+		}
+		for i, s := range signatureAlgos {
+			t.Run(fmt.Sprintf("case %v: %v", i, s), func(t *testing.T) {
+				key := make([]byte, 256)
+				rand.Read(key)
```
Comment (tarakby) [inline fvm/crypto/crypto_test.go:271]: We can make the test less probabilistic this way (we make sure it fails most of the time) 

```suggestion
				key, err := crypto.ValidatePublicKey(s, validPublicKey(t, s))
				key[0] ^= 1  // alter one bit of the valid key 
```

File: fvm/crypto/crypto.go (hunk 1 above)
Comment (janezpodhostnik) [reply fvm/crypto/crypto.go:160]: It was returning false when the condition I use now returns true.

**Review by tarakby — APPROVED** [review]
Thanks Janez! 🏄

File: fvm/crypto/crypto_test.go [hunk 5]
```diff
@@ -227,3 +227,53 @@ func TestVerifySignatureFromRuntime(t *testing.T) {
 		}
 	})
 }
+
+func TestValidatePublicKey(t *testing.T) {
+
+	validPublicKey := func(t *testing.T, s runtime.SignatureAlgorithm) []byte {
+		seed := make([]byte, 256)
```
Comment (turbolent) [inline fvm/crypto/crypto_test.go:234]: Maybe use an existing constant or refactor this magic number to a descriptive constant. Why 256?

**Review by turbolent — APPROVED** [review]
👍

File: fvm/crypto/crypto.go (hunk 1 above)
Comment (tarakby) [reply fvm/crypto/crypto.go]: fixed [here](https://github.com/onflow/flow-go/pull/788/commits/80e5fc3c10bf744bd0a569705601da5264e21d2f)

**Pull request merged**